include/cuco/detail/bloom_filter/bloom_filter_policy_impl.cuh

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/hash_functions.cuh>
+
+// TODO switch to <cuda/std/algorithm> once available
+#include <cuda/std/__algorithm/max.h>
+#include <cuda/std/__algorithm/min.h>
+#include <cuda/std/array>
+#include <cuda/std/bit>
+#include <cuda/std/limits>
+#include <cuda/std/tuple>
+#include <cuda/std/type_traits>
+
+#include <cstdint>
+
+namespace cuco::detail {
+
+template <class Hash, class Block>
+class bloom_filter_policy_impl {
+ public:
+  using hasher             = Hash;
+  using word_type          = typename Block::value_type;
+  using hash_argument_type = typename hasher::argument_type;
+  using hash_result_type   = decltype(std::declval<hasher>()(std::declval<hash_argument_type>()));
+
+  static constexpr std::uint32_t words_per_block = cuda::std::tuple_size_v<Block>;
+
+ private:
+  static constexpr std::uint32_t word_bits =
+    cuda::std::numeric_limits<typename Block::value_type>::digits;
+  static constexpr std::uint32_t bit_index_width = cuda::std::bit_width(word_bits - 1);
+
+ public:
+  static_assert(cuda::std::has_single_bit(words_per_block) and words_per_block <= 32,
+                "Number of words per block must be a power-of-two and less than or equal to 32");
+
+  __host__ __device__ constexpr bloom_filter_policy_impl(uint32_t pattern_bits, Hash hash)
+    : hash_{hash}
+  {
+    constexpr uint32_t hash_bits = cuda::std::numeric_limits<hash_result_type>::digits;
+
+    // This ensures each word in the block has at least one bit set; otherwise we would never use
+    // some of the words
+    constexpr uint32_t min_pattern_bits = words_per_block;
+
+    // The maximum number of bits to be set for a key is capped by the number of bits in the hash
+    // value or the number of bits in the filter block
+    constexpr uint32_t max_pattern_bits =
+      cuda::std::min(hash_bits / bit_index_width, word_bits * words_per_block);
+
+    // TODO do we want this to fail?
+    // constexpr uint32_t max_pattern_bits_from_hash = hash_bits / bit_index_width;
+    // static_assert(min_pattern_bits <= max_pattern_bits_from_hash, "Hash value too narrow for the
+    // specified filter block type");
+
+    // TODO update this part once cuda::std::clamp is available
+    // Here we adjust the value dynamically; alternatively we could throw an error if the value is
```

Comment:
I think an error is the right result here. We wouldn't want, for instance, a caller to think they were setting some number bits sufficient to achieve a specific false positive probability and then have that be lowered slightly in this constructor.

Reply:
> think they were setting some number bits sufficient to achieve a specific false positive probability and then have that be lowered slightly in this constructor

Very good point! I updated the ctor so that it will throw an error now.